mono/utils/mono-threads.c

```diff
@@ -935,21 +983,45 @@ suspend_sync (MonoNativeThreadId tid, gboolean interrupt_kernel)
 	if (!info)
 		return NULL;
 
-	switch (mono_threads_transition_request_async_suspension (info)) {
-	case AsyncSuspendAlreadySuspended:
+	MonoRequestSuspendResult result;
+	switch ((result = mono_threads_transition_request_initiate_suspension (info))) {
+	case ReqSuspendAlreadySuspended:
 		mono_hazard_pointer_clear (hp, 1); //XXX this is questionable we got to clean the suspend/resume nonsense of critical sections
 		return info;
-	case AsyncSuspendInitSuspend:
-		if (!begin_async_suspend (info, interrupt_kernel)) {
+	case ReqSuspendInitSuspendRunning:
+		if (!begin_suspend_for_running_thread (info, interrupt_kernel)) {
 			mono_hazard_pointer_clear (hp, 1);
 			return NULL;
 		}
 		break;
-	case AsyncSuspendBlocking:
-		if (interrupt_kernel)
+	case ReqSuspendAlreadySuspendedBlocking:
+	case ReqSuspendInitSuspendBlocking: {
+		gboolean did_interrupt = FALSE;
+		if (result == ReqSuspendInitSuspendBlocking) {
```

Comment:
We should duplicate the code for `ReqSuspendAlreadySuspendedBlocking` and `ReqSuspendInitSuspendBlocking`